ec/src/models/bw6/mod.rs

```diff
@@ -114,23 +116,45 @@ pub trait BW6Config: 'static + Eq + Sized {
             })
             .product::<<BW6<Self> as Pairing>::TargetField>();
 
+        let f_u_inv;
+
+        // TODO: is it enough to get the inverse of f_1, or does f_u also need to get inverted?
         if Self::ATE_LOOP_COUNT_1_IS_NEGATIVE {
-            f_1.cyclotomic_inverse_in_place();
+            f_u_inv = f_u;
+            f_u.cyclotomic_inverse_in_place();
+        } else {
+            f_u_inv = f_u.cyclotomic_inverse().unwrap();
         }
+
+        // f_1(P) = f_(u+1)(P) = f_u(P) * l([u]q, q)(P)
+        let mut f_1 = cfg_chunks_mut!(pairs_1, 4)
+            .map(|pairs| {
+                let mut f = f_u;
+                for (p, coeffs) in pairs.iter_mut() {
+                    BW6::<Self>::ell(&mut f, &coeffs.next().unwrap(), &p.0);
+                }
+                f
+            })
+            .product::<<BW6<Self> as Pairing>::TargetField>();
+
         let mut f_2 = cfg_chunks_mut!(pairs_2, 4)
             .map(|pairs| {
-                let mut f = <<BW6<Self> as Pairing>::TargetField>::one();
+                let mut f = f_u;
                 for i in (1..Self::ATE_LOOP_COUNT_2.len()).rev() {
-                    if i != Self::ATE_LOOP_COUNT_2.len() - 1 {
-                        f.square_in_place();
-                    }
+                    f.square_in_place();
 
                     for (p, ref mut coeffs) in pairs.iter_mut() {
                         BW6::<Self>::ell(&mut f, &coeffs.next().unwrap(), &p.0);
                     }
 
                     let bit = Self::ATE_LOOP_COUNT_2[i - 1];
-                    if bit == 1 || bit == -1 {
+                    if bit == 1 {
+                        f *= &f_u;
+                        for &mut (p, ref mut coeffs) in pairs.iter_mut() {
+                            BW6::<Self>::ell(&mut f, &coeffs.next().unwrap(), &p.0);
+                        }
+                    } else if bit == -1 {
+                        f *= &f_u_inv;
```

Comment:
The for loops in both cases are the same, so they can be extracted to outside the loop, no?